### PR TITLE
Handle host deletions dynamically

### PR DESCRIPTION
**Pull request by Arun-LinkedIn**
Changes to remove host and its associated replicas from in-memory data structures when it is deleted in cluster.

### CODE REVIEW

**Review by justinlin-linkedin — APPROVED** [review]
LGTM

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreToDataNodeConfigAdapter.java [hunk 1]
```diff
@@ -47,6 +47,7 @@
  */
 class PropertyStoreToDataNodeConfigAdapter implements DataNodeConfigSource {
   static final String CONFIG_PATH = "/DataNodeConfigs";
+  static final String SEPARATOR = "/";
```
Comment (snalli) [inline ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreToDataNodeConfigAdapter.java:50]: File.separator

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreToDataNodeConfigAdapter.java [hunk 2]
```diff
@@ -105,6 +106,25 @@ public List<String> getAllDataNodeNames() {
...
+  }
+
+  /**
+   * Get the name of instance
+   * @param path path to ZNode
+   * @return the name of instance
+   */
+  private String getInstanceName(String path) {
+    // Instance name is the path after /DataNodeConfigs/
+    return path.substring(CONFIG_PATH.length() + SEPARATOR.length());
```
Comment (snalli) [inline ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreToDataNodeConfigAdapter.java:125]: path.substr(last index of separator)

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreToDataNodeConfigAdapter.java [hunk 3]
```diff
@@ -105,6 +106,25 @@ public List<String> getAllDataNodeNames() {
     return names == null ? Collections.emptyList() : names;
   }
 
+  /**
+   * Get path to ZNode
+   * @param instanceName the name of instance
+   * @return path to ZNode
+   */
+  private String getPath(String instanceName) {
+    return CONFIG_PATH + SEPARATOR + instanceName;
```
Comment (snalli) [inline ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreToDataNodeConfigAdapter.java:115]: does not matter here but u may consider string.join if the path grows

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreToDataNodeConfigAdapter.java [hunk 4]
```diff
@@ -105,6 +106,25 @@ public List<String> getAllDataNodeNames() {
     return names == null ? Collections.emptyList() : names;
   }
 
+  /**
+   * Get path to ZNode
+   * @param instanceName the name of instance
+   * @return path to ZNode
+   */
+  private String getPath(String instanceName) {
```
Comment (snalli) [inline ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreToDataNodeConfigAdapter.java:114]: i feel private is too strong for this, public or protected would be better

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreToDataNodeConfigAdapter.java [hunk 5]
```diff
@@ -105,6 +106,25 @@ public List<String> getAllDataNodeNames() {
...
+    return CONFIG_PATH + SEPARATOR + instanceName;
+  }
+
+  /**
+   * Get the name of instance
+   * @param path path to ZNode
+   * @return the name of instance
+   */
+  private String getInstanceName(String path) {
```
Comment (snalli) [inline ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreToDataNodeConfigAdapter.java:123]: public or protected

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreToDataNodeConfigAdapter.java [hunk 6]
```diff
@@ -142,14 +162,19 @@ public void onDataCreate(String path) {
 
     @Override
     public void onDataDelete(String path) {
-      // TODO handle node deletions dynamically. Doing so requires further work in the ClusterChangeHandler impl
-      LOGGER.info("DataNodeConfig path {} deleted. This requires a restart to handle", path);
+      // Handle node deletions dynamically.
+      LOGGER.info("DataNodeConfig path {} deleted", path);
+      onPathDelete(path);
```
Comment (snalli) [inline ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreToDataNodeConfigAdapter.java:167]: single usage one-liner fn can be replaced with the contents of the fn

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixClusterManager.java [hunk 7]
```diff
@@ -653,6 +654,15 @@ private void addClusterWideRawCapacity(long diskRawCapacityBytes) {
     clusterWideRawCapacityBytes.getAndAdd(diskRawCapacityBytes);
   }
 
+  /**
+   * Reduce cluster-wide raw capacity. This is called when new disk is removed from cluster. We update
+   * cluster-wide raw capacity by removing the disk capacity.
+   * @param diskRawCapacityBytes raw disk capacity to be decreased.
+   */
+  private void reduceClusterWideRawCapacity(long diskRawCapacityBytes) {
+    clusterWideRawCapacityBytes.getAndAdd(-diskRawCapacityBytes);
```
Comment (snalli) [inline ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixClusterManager.java:663]: single usage one liner. Replace its call site with this one liner itself.

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixClusterManager.java [hunk 8]
```diff
@@ -1004,6 +1019,21 @@ void handleDataNodeConfigChange(Iterable<DataNodeConfig> configs, String dcName,
       }
     }
 
+    /**
+     * Handle deletion of an instance from cluster.
+     * @param instanceName instance that was deleted
+     */
+    void handleDataNodeDelete(String instanceName) {
+      synchronized (notificationLock) {
+        deleteInstanceInfo(instanceName);
```
Comment (snalli) [inline ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixClusterManager.java:1028]: collapse deleteInstanceInfo here. So far i found only one usage of it and it looks like it must always be called from within this sync block. So don't open a path for someone to call it outside this sync block.

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixClusterManager.java [hunk 9]
```diff
@@ -1351,6 +1402,21 @@ private List<ReplicaId> createNewInstance(DataNodeConfig dataNodeConfig, String
       return addedReplicas;
     }
 
+    /**
+     * Delete an instance(node) and remove all disks/replicas associated with it.
+     * @param instanceName the deleted instance
+     * @return a list of replicas present on this instance;
+     */
+    private List<ReplicaId> deleteInstance(String instanceName) {
```
Comment (snalli) [inline ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixClusterManager.java:1410]: this can be folded into deleteInstanceInfo

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixClusterManager.java [hunk 10]
```diff
@@ -1408,6 +1474,27 @@ private List<ReplicaId> initializeDisksAndReplicasOnNode(AmbryDataNode datanode,
       return addedReplicas;
     }
 
+    /**
+     * Remove the disks and replicas on the given node. Remove associated partitions if this is the only partition.
+     * @param datanode the {@link AmbryDataNode} that is being initialized.
+     * @return a list of replicas present on this deleted node.
+     */
+    private List<ReplicaId> removeDisksAndReplicasOnNode(AmbryDataNode datanode) {
```
Comment (snalli) [inline ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixClusterManager.java:1482]: this can also be folded in the caller. Write comments because the top caller will be big when u combine all these into it.

**Comment by snalli** [comment]
Overall lgtm, i understand the flow but the code is fragmented into mutliple small or one-liner singe-usage fns and that was a bit hard to follow.

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixClusterManager.java (hunk 7 above)
Comment (Arun-LinkedIn) [reply ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixClusterManager.java]: Done

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixClusterManager.java (hunk 8 above)
Comment (Arun-LinkedIn) [reply ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixClusterManager.java]: Yeah, agree with you. Updated to inline the function.

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixClusterManager.java (hunk 9 above)
Comment (Arun-LinkedIn) [reply ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixClusterManager.java]: Done

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixClusterManager.java (hunk 10 above)
Comment (Arun-LinkedIn) [reply ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixClusterManager.java]: Done

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreToDataNodeConfigAdapter.java (hunk 1 above)
Comment (Arun-LinkedIn) [reply ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreToDataNodeConfigAdapter.java]: Since the is a path separator used in zookeeper nodes and is always `/ `, I am wondering if it is better to not use File.separator (which could be `\\` on windows systems although we use only linux).

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreToDataNodeConfigAdapter.java (hunk 4 above)
Comment (Arun-LinkedIn) [reply ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreToDataNodeConfigAdapter.java]: Sure. Agree.

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreToDataNodeConfigAdapter.java (hunk 3 above)
Comment (Arun-LinkedIn) [reply ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreToDataNodeConfigAdapter.java]: Yeah, true. Left it for now since path is small.

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreToDataNodeConfigAdapter.java (hunk 5 above)
Comment (Arun-LinkedIn) [reply ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreToDataNodeConfigAdapter.java]: Done

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreToDataNodeConfigAdapter.java (hunk 2 above)
Comment (Arun-LinkedIn) [reply ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreToDataNodeConfigAdapter.java]: Done

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreToDataNodeConfigAdapter.java (hunk 6 above)
Comment (Arun-LinkedIn) [reply ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreToDataNodeConfigAdapter.java:167]: In this case, I felt it is better to not inline it since it matches the pattern like you said.

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreToDataNodeConfigAdapter.java (hunk 1 above)
Comment (snalli) [reply ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreToDataNodeConfigAdapter.java]: ok

**Comment by snalli** [comment]
lgtm, u may merge when ready.

**Review by snalli — APPROVED** [review]
